<a href="https://colab.research.google.com/github/ganeshbmc/graded/blob/master/BA/GA2/ga2_ba.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from statsmodels.formula.api import ols

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

print('Imports successful!')

Imports successful!


In [10]:
df = pd.read_excel(f"data_transform.xlsx")
df.head()

Sales  Price
0     73    1.5
1     32    3.9
2     35    3.8
3     47    1.9
4     28    4.2

In [11]:
df.shape

(50, 2)

In [12]:
df.describe()

Sales      Price
count   50.0000  50.000000
mean    51.5600   2.784000
std     28.7673   0.984235
min     26.0000   1.300000
25%     34.2500   1.825000
50%     40.0000   2.850000
75%     60.7500   3.775000
max    165.0000   4.200000

`Q1`  

In [13]:
model_ols = ols('Sales ~ Price', data=df).fit()
model_ols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.526
Model:                            OLS   Adj. R-squared:                  0.516
Method:                 Least Squares   F-statistic:                     53.34
Date:                Thu, 20 Jul 2023   Prob (F-statistic):           2.52e-09
Time:                        17:48:09   Log-Likelihood:                -219.72
No. Observations:                  50   AIC:                             443.4
Df Residuals:                      48   BIC:                             447.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    110.5953      8.564     12.914      0.000      93.377     127.814
Price        -21.2052      2.903     -7.304      0.000     -27.043     -15.368
==============================================================================
Omnibus:                       35.987   Durbin-Watson:                   2.229
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              104.127
Skew:                           1.955   Prob(JB):                     2.45e-23
Kurtosis:                       8.890   Cond. No.                         9.85
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

`Ans: 0.53`  

`Q2`  

`Ans: Yes (p=2.52e-09)`  

`Q3`  

`Ans: 110.60`  

`Q4`  

`Ans: -21.21`  

## Transformed dataset  

In [14]:
log_df = df.copy()
log_df['log_sales'] = np.log(log_df['Sales'])
log_df['log_price'] = np.log(log_df['Price'])
# log_df.drop(['Sales', 'Price'], axis=1, inplace=True)
log_df.head()

Sales  Price  log_sales  log_price
0     73    1.5   4.290459   0.405465
1     32    3.9   3.465736   1.360977
2     35    3.8   3.555348   1.335001
3     47    1.9   3.850148   0.641854
4     28    4.2   3.332205   1.435085

In [15]:
model_log_ols = ols('log_sales ~ log_price', data=log_df).fit()
model_log_ols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              log_sales   R-squared:                       0.734
Model:                            OLS   Adj. R-squared:                  0.728
Method:                 Least Squares   F-statistic:                     132.4
Date:                Thu, 20 Jul 2023   Prob (F-statistic):           2.10e-15
Time:                        17:48:09   Log-Likelihood:                 3.4431
No. Observations:                  50   AIC:                            -2.886
Df Residuals:                      48   BIC:                            0.9379
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.7793      0.088     54.030      0.000       4.601       4.957
log_price     -0.9898      0.086    -11.507      0.000      -1.163      -0.817
==============================================================================
Omnibus:                        4.257   Durbin-Watson:                   2.148
Prob(Omnibus):                  0.119   Jarque-Bera (JB):                3.136
Skew:                           0.521   Prob(JB):                        0.209
Kurtosis:                       3.647   Cond. No.                         5.24
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

`Q5`  

`Ans: 0.73`  

`Q6`  

`Ans: Yes (p=2.10e-15)`  

`Q7`  

`Ans: 4.78`  

`Q8`  

`Ans: -0.99`  

## Car price prediction  

In [16]:
X = pd.read_csv(f"X.csv").iloc[:, 1:]
y = pd.read_csv(f"y.csv").iloc[:, 1:]
X.shape, y.shape

((205, 10), (205, 1))

In [17]:
X.head()

wheelbase  boreratio  compressionratio  fueltype_diesel  \
0       88.6       3.47               9.0                0   
1       88.6       3.47               9.0                0   
2       94.5       2.68               9.0                0   
3       99.8       3.19              10.0                0   
4       99.4       3.19               8.0                0   

   carbody_convertible  enginelocation_rear  cylindernumber_four  \
0                    1                    0                    1   
1                    1                    0                    1   
2                    0                    0                    0   
3                    0                    0                    1   
4                    0                    0                    0   

   cylindernumber_twelve  fuelsystem_idi  CarCompany_high  
0                      0               0                0  
1                      0               0                0  
2                      0               0                0  
3                      0               0                0  
4                      0               0                0

In [18]:
y.head()

price
0  13495.0
1  16500.0
2  16500.0
3  13950.0
4  17450.0

In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((164, 10), (164, 1), (41, 10), (41, 1))

In [20]:
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression()

In [21]:
y_pred = lin_reg.predict(X_test)
y_pred

array([[ 7055.12887962],
       [18742.75463112],
       [17062.0266149 ],
       [10119.18680525],
       [ 8412.45788306],
       [10691.05211314],
       [ 6787.60626307],
       [ 8304.02750192],
       [13257.98904804],
       [ 6832.74710105],
       [25290.7097788 ],
       [24606.32900823],
       [14456.85590676],
       [12884.71370777],
       [ 7111.22698042],
       [10691.05211314],
       [ 9816.70872294],
       [19827.73827993],
       [ 8924.90689453],
       [ 8304.02750192],
       [10356.22662574],
       [15550.55897468],
       [ 9816.70872294],
       [10691.05211314],
       [17907.02575028],
       [ 8819.19528605],
       [ 8656.99247411],
       [18454.55622296],
       [ 8656.99247411],
       [ 7610.1799255 ],
       [ 9394.51558478],
       [10823.46661171],
       [17275.54295799],
       [12342.5292153 ],
       [ 7610.1799255 ],
       [28114.52224767],
       [16761.88027934],
       [11106.64151476],
       [ 6787.60626307],
       [32486.42588023],


`Q9`  

In [22]:
lin_reg.score(X_test, y_test)

0.7914295570653098

In [23]:
r2_score(y_test, y_pred)

0.7914295570653098

In [24]:
lin_reg.score(X_train, y_train)

0.9065828499654524

`Q10`  

In [25]:
lin_reg.intercept_

array([-28636.04595786])

`Q11`  

In [26]:
test_data = np.array([88.46, 4, 9, 1, 0, 1, 0, 1, 0, 1])
len(test_data)

10

In [27]:
prediction = lin_reg.predict([test_data])
prediction

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[50762.92866796]])